# An Introduction to Python, part 1 (lessons 1-7)

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#1-Values" data-toc-modified-id="1-Values-1">1 Values</a></span><ul class="toc-item"><li><span><a href="#1.1-Integers" data-toc-modified-id="1.1-Integers-1.1">1.1 Integers</a></span></li><li><span><a href="#1.2-Floats" data-toc-modified-id="1.2-Floats-1.2">1.2 Floats</a></span></li><li><span><a href="#1.3-String-values" data-toc-modified-id="1.3-String-values-1.3">1.3 String values</a></span></li><li><span><a href="#1.4-Boolean-values" data-toc-modified-id="1.4-Boolean-values-1.4">1.4 Boolean values</a></span></li><li><span><a href="#1.5-The-None-value" data-toc-modified-id="1.5-The-None-value-1.5">1.5 The None value</a></span></li></ul></li><li><span><a href="#2-Expressions-and-operators" data-toc-modified-id="2-Expressions-and-operators-2">2 Expressions and operators</a></span><ul class="toc-item"><li><span><a href="#2.1-Order-of-operations" data-toc-modified-id="2.1-Order-of-operations-2.1">2.1 Order of operations</a></span><ul class="toc-item"><li><span><a href="#2.1.1-(Optional)-Deep-dive-into-Order-of-Operations" data-toc-modified-id="2.1.1-(Optional)-Deep-dive-into-Order-of-Operations-2.1.1">2.1.1 (Optional) Deep dive into Order of Operations</a></span></li><li><span><a href="#2.1.2-Exercise-–-where-to-put-the-parenthesis?" data-toc-modified-id="2.1.2-Exercise-–-where-to-put-the-parenthesis?-2.1.2">2.1.2 Exercise – where to put the parenthesis?</a></span></li></ul></li><li><span><a href="#2.2-Concatination" data-toc-modified-id="2.2-Concatination-2.2">2.2 Concatination</a></span></li></ul></li><li><span><a href="#3-Variables-and-statements" data-toc-modified-id="3-Variables-and-statements-3">3 Variables and statements</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#3.0.1-Quick-sidenote-–-the-print()-function" data-toc-modified-id="3.0.1-Quick-sidenote-–-the-print()-function-3.0.1">3.0.1 Quick sidenote – the <code>print()</code> function</a></span></li></ul></li><li><span><a href="#3.1-Assigning-variables" data-toc-modified-id="3.1-Assigning-variables-3.1">3.1 Assigning variables</a></span></li><li><span><a href="#3.2-Variables-and-strings" data-toc-modified-id="3.2-Variables-and-strings-3.2">3.2 Variables and strings</a></span><ul class="toc-item"><li><span><a href="#3.2.1-Long-strings" data-toc-modified-id="3.2.1-Long-strings-3.2.1">3.2.1 Long strings</a></span></li><li><span><a href="#3.2.2-String-formatting-–-inserting-variables-into-string-values" data-toc-modified-id="3.2.2-String-formatting-–-inserting-variables-into-string-values-3.2.2">3.2.2 String formatting – inserting variables into string values</a></span></li><li><span><a href="#3.2.3-(Deepdive)-–-Long-strings,-string-literals-and-the-escape-character-\" data-toc-modified-id="3.2.3-(Deepdive)-–-Long-strings,-string-literals-and-the-escape-character-\-3.2.3">3.2.3 (Deepdive) – Long strings, string literals and the escape character <code>\</code></a></span><ul class="toc-item"><li><span><a href="#3.2.3.1-More-on-tripple-quote-strings" data-toc-modified-id="3.2.3.1-More-on-tripple-quote-strings-3.2.3.1">3.2.3.1 More on tripple quote strings</a></span></li><li><span><a href="#3.2.3.2-String-literals-and-the-backslash-\" data-toc-modified-id="3.2.3.2-String-literals-and-the-backslash-\-3.2.3.2">3.2.3.2 String literals and the backslash <code>\</code></a></span></li><li><span><a href="#3.2.3.3-Exercise-–-write-a-letter!" data-toc-modified-id="3.2.3.3-Exercise-–-write-a-letter!-3.2.3.3">3.2.3.3 Exercise – write a letter!</a></span></li></ul></li></ul></li><li><span><a href="#3.3-Changing-variables-–-Operators-and-assignment-statements" data-toc-modified-id="3.3-Changing-variables-–-Operators-and-assignment-statements-3.3">3.3 Changing variables – Operators and assignment statements</a></span><ul class="toc-item"><li><span><a href="#3.3.1-Augmented-assignment-operators" data-toc-modified-id="3.3.1-Augmented-assignment-operators-3.3.1">3.3.1 Augmented assignment operators</a></span></li></ul></li><li><span><a href="#3.4-Multiple-assignment-statements" data-toc-modified-id="3.4-Multiple-assignment-statements-3.4">3.4 Multiple assignment statements</a></span></li><li><span><a href="#3.5-(Optional)-More-about-readability-–-the-zen-of-Python" data-toc-modified-id="3.5-(Optional)-More-about-readability-–-the-zen-of-Python-3.5">3.5 (Optional) More about readability – the zen of Python</a></span></li><li><span><a href="#3.6-Exercise-–-string-concatination-and-variables" data-toc-modified-id="3.6-Exercise-–-string-concatination-and-variables-3.6">3.6 Exercise – string concatination and variables</a></span></li></ul></li><li><span><a href="#4-Comparing-values" data-toc-modified-id="4-Comparing-values-4">4 Comparing values</a></span><ul class="toc-item"><li><span><a href="#4.1-Comparison-operators" data-toc-modified-id="4.1-Comparison-operators-4.1">4.1 Comparison operators</a></span></li><li><span><a href="#4.2-Boolean-operators" data-toc-modified-id="4.2-Boolean-operators-4.2">4.2 Boolean operators</a></span><ul class="toc-item"><li><span><a href="#4.2.1-The-and-operator" data-toc-modified-id="4.2.1-The-and-operator-4.2.1">4.2.1 The <code>and</code> operator</a></span></li><li><span><a href="#4.2.2-The-or-operator" data-toc-modified-id="4.2.2-The-or-operator-4.2.2">4.2.2 The <code>or</code> operator</a></span></li><li><span><a href="#4.2.3-The-not-operator" data-toc-modified-id="4.2.3-The-not-operator-4.2.3">4.2.3 The <code>not</code> operator</a></span></li><li><span><a href="#4.2.4-Boolean-operators-–-examples" data-toc-modified-id="4.2.4-Boolean-operators-–-examples-4.2.4">4.2.4 Boolean operators – examples</a></span></li></ul></li><li><span><a href="#4.3-Exercise-–-insert-the-correct-operators" data-toc-modified-id="4.3-Exercise-–-insert-the-correct-operators-4.3">4.3 Exercise – insert the correct operators</a></span></li></ul></li><li><span><a href="#5-Flow-Control-Statements" data-toc-modified-id="5-Flow-Control-Statements-5">5 Flow Control Statements</a></span><ul class="toc-item"><li><span><a href="#5.1-The-if-statement" data-toc-modified-id="5.1-The-if-statement-5.1">5.1 The <code>if</code>-statement</a></span></li><li><span><a href="#5.2-The-else-statement" data-toc-modified-id="5.2-The-else-statement-5.2">5.2 The <code>else</code>-statement</a></span></li><li><span><a href="#5.3-The-elif-statement" data-toc-modified-id="5.3-The-elif-statement-5.3">5.3 The <code>elif</code>-statement</a></span></li><li><span><a href="#5.4-(Optional)-&quot;Truthy&quot;-and-&quot;Falsy&quot;-values" data-toc-modified-id="5.4-(Optional)-&quot;Truthy&quot;-and-&quot;Falsy&quot;-values-5.4">5.4 (Optional) "Truthy" and "Falsy" values</a></span></li><li><span><a href="#5.5-Exercise-–-type-the-end-of-the-story!" data-toc-modified-id="5.5-Exercise-–-type-the-end-of-the-story!-5.5">5.5 Exercise – type the end of the story!</a></span></li></ul></li><li><span><a href="#6-Functions" data-toc-modified-id="6-Functions-6">6 Functions</a></span><ul class="toc-item"><li><span><a href="#6.1-Definitions-and-terminology" data-toc-modified-id="6.1-Definitions-and-terminology-6.1">6.1 Definitions and terminology</a></span></li><li><span><a href="#6.2-Built-in-Functions" data-toc-modified-id="6.2-Built-in-Functions-6.2">6.2 Built-in Functions</a></span></li><li><span><a href="#6.3-The-standard-library-and-modules" data-toc-modified-id="6.3-The-standard-library-and-modules-6.3">6.3 The standard library and modules</a></span><ul class="toc-item"><li><span><a href="#6.3.1-Exercise-–-the-coin-toss" data-toc-modified-id="6.3.1-Exercise-–-the-coin-toss-6.3.1">6.3.1 Exercise – the coin toss</a></span></li></ul></li><li><span><a href="#6.4-Third-party-modules" data-toc-modified-id="6.4-Third-party-modules-6.4">6.4 Third party modules</a></span></li><li><span><a href="#6.5-Writing-your-own-functions" data-toc-modified-id="6.5-Writing-your-own-functions-6.5">6.5 Writing your own functions</a></span></li><li><span><a href="#6.6-Arguments" data-toc-modified-id="6.6-Arguments-6.6">6.6 Arguments</a></span><ul class="toc-item"><li><span><a href="#6.6.1-Exercise-–-truth-teller-function" data-toc-modified-id="6.6.1-Exercise-–-truth-teller-function-6.6.1">6.6.1 Exercise – truth-teller function</a></span></li></ul></li><li><span><a href="#6.7-(Optional)-Styling-conventions-writing-functions" data-toc-modified-id="6.7-(Optional)-Styling-conventions-writing-functions-6.7">6.7 (Optional) Styling conventions writing functions</a></span></li><li><span><a href="#6.8-The-return-statement" data-toc-modified-id="6.8-The-return-statement-6.8">6.8 The <code>return</code> statement</a></span></li><li><span><a href="#6.9-Exercise-–-the-currency-converter" data-toc-modified-id="6.9-Exercise-–-the-currency-converter-6.9">6.9 Exercise – the currency converter</a></span></li><li><span><a href="#6.10-Bulletpoints" data-toc-modified-id="6.10-Bulletpoints-6.10">6.10 Bulletpoints</a></span></li></ul></li><li><span><a href="#7-Local-and-global-scopes" data-toc-modified-id="7-Local-and-global-scopes-7">7 Local and global scopes</a></span><ul class="toc-item"><li><span><a href="#7.1-Global-code-cannot-use-local-variables" data-toc-modified-id="7.1-Global-code-cannot-use-local-variables-7.1">7.1 Global code cannot use local variables</a></span></li><li><span><a href="#7.2-Global-variables-can-be-used-in-local-scopes" data-toc-modified-id="7.2-Global-variables-can-be-used-in-local-scopes-7.2">7.2 Global variables can be used in local scopes</a></span></li><li><span><a href="#7.3-Local-scopes-can't-use-variables-in-other-local-scopes" data-toc-modified-id="7.3-Local-scopes-can't-use-variables-in-other-local-scopes-7.3">7.3 Local scopes can't use variables in other local scopes</a></span></li><li><span><a href="#7.4-Exercise-–-why-won't-the-code-work?" data-toc-modified-id="7.4-Exercise-–-why-won't-the-code-work?-7.4">7.4 Exercise – why won't the code work?</a></span></li></ul></li></ul></div>

**Introduction**

Python is a programming language. "Language" is a pretty good analogy since python, just like, say, French, consists of some basic foundations – there is the basic "grammar" and the vocabulary. If you learn the basics of a language, you will hopefully know enough to experiment with it! Here, we will therefore start to look at the basic stuff in Python. 

The analogy of learning a language also reflects another important aspect. Learning programming, just as learning a language, from scratch requires time and practice. This notebook will (hopefully) give you the foundation to stand on and enough guidance to start experimenting. But to fully learn Python, you will need to invest time.

**Words of wisdom**

You don't need to be a professional programmer to solve various problems with programming. Often, some simple Python code will do the trick. Going through this notebook, you may sometimes get completely stuck. In those times I recommend that you continue try this procedure:

1. Remember that it isn't learning rocket science, and it isn't learning math. You are not stupid!
2. If stuck, try reading again and experiment with what you've read
3. Google it! Even though I find the company Google terrifying, their search engine is a fantastic way to help you move forward when stuck! Try to put you problem into words and Google it!

Also, going through this notebook, you might think there is a little to much emphasis on definitions of various Python concepts. But this is so that when you google stuff to find out why code you wrote didn't work, you'll be familiar with [Python terminology](https://www.w3schools.com/python/python_ref_glossary.asp). You will hopefully have som idea of which concept you want to know more about. One aim with this notebook is to be a reference to return to if you need a quick update on a particular subject.


**Working in a Jupyter Notebook**

Jupyter Notebooks are great for learning. It is really easy to experiment with code. I encourage you to try out the concepts you're about to read about. Here is a Jupyter Notebook cheat-sheet for hotkeys to keep in mind working through this course:

- The keys `A` and `B` creates new cells in your notebook. `A` creates a cell above the cell currently selected, and `B` creates one below. Each newly created cell is by default a code cell
- `M` sets a cell into a markdown (text input) cell
- `Y` sets a cell into a code input cell
- <kbd>Shift ⇧</kbd>+<kbd>Enter ↵</kbd> runs a cell, showing you the output of your code (or markdown)
- `D,D` (that is, the `D` key typed twice swiftly) will remove the selected cell
- `Z` will undo deleted cells (note that this won't work if you're editing within a cell)


Don't hesitate to add some extra cells and try out the things you read about. That's the best way to learn!

**About this course (full disclosure)**

This notebook's content has been fully created by me – Johan Ekman. I am working full time as a data journalist at one of Sweden's biggest news papers, and Python is one of my main tools in every day work. But I don't consider myself a professional programmer. This notebook is based on my knowledge of the language, and there may be some explanation that is not 100 per cent accurate. If you spot any errors or misleading explanations, please let me know at johan.h.ekman@gmail.com.

Also, a big inspiration was Al Sweigart's course [Automate the boring stuff with Python](https://automatetheboringstuff.com/). I highly recommend it!

---

## 1 Values

The most basic structures in Python are values. All programming aims to do something with some values. There are five basic value types:

- Integers
- Floating points (i.e floats)
- Strings
- Booleans
- Nonetype (or just the `None` value)

### 1.1 Integers

Integers are just plain numbers:

In [2]:
10

10

We can always check what type of value we have with the `type()` function (more on functions in chapter 6):

In [3]:
type(10)

int

`int` for integer! We can do negative numbers as well:

In [2]:
-10

-10

Or we can do as large numbers we'd like:

In [3]:
3243667746

3243667746

In Python, there is a constant quest to make the language more readable (more about this later in the course). Therefore, here is a useful trick for typing really big integers: you can include an underscore `_` to make the integer more readable. So instead of:

In [4]:
3243667746

3243667746

...we can type:

In [5]:
3_243_667_746

3243667746

This makes it easier to see that the number is around three billion!

### 1.2 Floats

In [4]:
10.5 # float

10.5

In [5]:
type(10.5)

float

 A float is simply a decimal number. "Float" is short for "floating point number".

### 1.3 String values

In [6]:
"Hello world!"

'Hello world!'

In [7]:
type("Hello world!")

str

String values can be typed with either single quotation marks `''`, or double `""`:

In [4]:
type('Hello world!')

str

### 1.4 Boolean values

Booleans indicates "truthiness". Simply if something is `True` or `False`.

In [8]:
True

True

In [10]:
False

False

In [9]:
type(True)

bool

Please note that Python is a literal programming language, the syntax is important. The value `False` is spelled like so, with a capital `F`. Spelling it `false` will produce an error:

In [13]:
False

False

In [14]:
false

NameError: name 'false' is not defined

### 1.5 The None value

In programming, there are also values to indicate non-values, or "nothingness". In Python, the most common one is the `None` value, or the `NoneType`.

In [143]:
None

Please note that running above cell produces no output. That is because this value indicates no input. There is simply nothing to return from a non-value.

The reason why such a valuetype is important is rather abstract. Let's just say that it is sometimes crucial in programming to be able to signal that nothing exists, or has happend in your code's execution.

In [18]:
type(None)

NoneType

---

## 2 Expressions and operators

One of Python's strengths is that it is both easy to read and to write. This is because we use words in our code that makes it easier for us humans to read and understand what is going on. Instead of, you know, only reading ones and zeros and not understanding anything. 

But the computer only uses ones and zeros, it needs our code in a language that it can process. This means that our Python language has to to be translated into a language that the computer then can read. This process is refered to as the interpretation of our code. And **the interpreter** is the computer reading our Python code and then translating it to "machine language", before executing it.

The most basic form of "instruction" you can give your computer – you can give the interpreter – is an **expression**. Expressions always evaluates, they always reduce, into a single value. This means that giving the interpreter a single value is an expression:

In [144]:
3 # this is an expression

3

But we can also let values interact with eachother in expressions, using **operators**. The most basic ones are the arithmetic operators (basically, math operations). For example:

In [3]:
1 + 1

2

In the cell above, we have two integer values that we add together through the _addition operator_ `+`. Here is another one, the subtraction operator, `-`:

In [20]:
1235 - 332

903

As you can see, we get an answer to the expression when we run the code. In this case, the integer 903. In Python, we say that the expression `1235 - 332` _returns_ the value 903.

More examples of operators:

In [22]:
5 * 3 # multiplication

15

In [23]:
123 / 7 # division

17.571428571428573

The division operator always returns a float value. Even if the result is even:

In [82]:
9 / 3

3.0

(Above, you see examples of **comments** in Python code. Everything that follows the `#` sign will be completely ignored by the interpreter. This is how you write comments in your code – to explain stuff without ruining your output! But it is also used to just exclude parts in your code for whatever reason)

We can evaluate longer expressions as well, containing more values:

In [4]:
2 + 4 + 3 + 41

50

And we can also mix them:

In [5]:
2 + 4 - 12

-6

### 2.1 Order of operations

In Python, code is read by the interpreter in a specific manner. Before the code is run, the interpreter reads the code. It starts at the top of the code to the left, and then reads line by line to the right – one line at the time. But it will then run the code in a specific order depending on certain factors. For example, some operators take precendence over others. As an example:

In [6]:
1 + 3 * 5

16

Some of you maybe expected the answer 20 (1 plus 3 equals 4, 4 times 5 equals 20). But since the multiplication operator takes precendence, the interpreter returns the value 16.

Above, the code is read from left to right. The python interpreter sees the operators and then reads the operations in the expression a specific order. This is called the _order of operations_. This order is:

1. Paranthesis
2. Multiplication/division
3. addition/subtraction

For example:

In [7]:
(1 + 3) * 5

20

Here, the interpreter reads the code, registeres the operators `()`, `+` and `*`. It then prioritises the code within the parenthesis (since code within parenthesis takes precedence), and then runs the resulting value `(1 + 3)` in the remaing operation `4 * 5`. 

**The Order of operations**

```

1. () Parenthesis
2. ** Exponant
3. *  Multiplication
4. /  Division
5. +  Addition
6. -  Subtraction

```

Note that there are a whole bunch more of operators that also take precendence above themselves and above ones. But these will suffice for now. 

#### 2.1.1 (Optional) Deep dive into Order of Operations

Let's experiment! (Remember that you can create your own cells in this notebook and try out on your own!)

Take this expression, contains four values and three operators (`+`, `/` and `**`). If we run them as is:

In [34]:
3 + 6 / 3 ** 5

3.0246913580246915

Let us now experiment with paranthesis at various places in the expression. Try to guess what will happen before you run the code:

In [35]:
(3 + 6) / 3 ** 5

0.037037037037037035

Above, `(3 + 6)` is run first, which equals 9. Then `3 ** 5`, which equals 243. And finally, `9 / 243`

In [37]:
9 / 243

0.037037037037037035

In [39]:
3 + (6 / 3) ** 5

35.0

Here, first `(6 / 3)`, which equals 2. Then, `2 ** 5`, equals 32. And finally, `3 + 32`, which is 35.

In [42]:
((3 + 6) / 3) ** 5

243.0

Here, there are parenthesis within parenthesis, also called nested parenthesis. Remember, the Python interpreter reads  the code from left to right. So it starts reading the parenthesis with its `(` furthest to the left. 

That is, the expression `(3 + 6) / 3`. 

Then it reads the encased parenthesis first (since it takes precedence over the `/` operator), `3 + 6` which is 9. 

Now, it divides the resulting value 9 with 3, equals 3.

Finally, it exponates 3 with 5 (`3 ** 5`), which is 243.

#### 2.1.2 Exercise – where to put the parenthesis?

Let's do a little exercise. I said in the introduction to this course that programming in Python isn't math. But here we have a small math exercise (sorry!). The goal here is to get you experimenting with the order of operations. We have this long chain of operations on values:

In [56]:
10 + 5 * 10 - 2 / 15 + 5 / 4 

61.11666666666667

With no parenthesis, this expression evaluates to ~61.12. Try to include parenthesis to **get the answer 1.5 instead**. You will find the solution in the solutions.ipynb notebook. Use these cells below to experiment your way through:

### 2.2 Concatination

In Python, some operators can also be used in expressions with string values. Adding string values together by using the `+` operator is called _concatination_.

In [145]:
"Hello " + "world!"

'Hello world!'

In [44]:
"H" + "ello world!"

'Hello world!'

You can add two string values together, but you cant subtract them. This will lead to an error:

In [45]:
"Hello world!" - "world!"

TypeError: unsupported operand type(s) for -: 'str' and 'str'

You can also use the multiplication operator on string values. This is called string replication (and is more of a curiosity than important to be honest; not widely used...)

In [46]:
"Hello world! " * 10

'Hello world! Hello world! Hello world! Hello world! Hello world! Hello world! Hello world! Hello world! Hello world! Hello world! '

Note that in expressions with string values, order of operations is still in play:

In [47]:
"Hello world" + "!" * 10

'Hello world!!!!!!!!!!'

In [48]:
("Hello world" + "! ") * 10

'Hello world! Hello world! Hello world! Hello world! Hello world! Hello world! Hello world! Hello world! Hello world! Hello world! '

## 3 Variables and statements

#### 3.0.1 Quick sidenote – the `print()` function

Below, we'll use the `print()` function. It takes any value within its parenthesis and "prints" it back to you. It is one (of many) of Python's _built-in functions_ (we will go through this thoroughly in chapter 6). In a Jupyter Notebook, this function is sort of redundant. In a notebook, the value being returned is always shown as the cell output when we run the code. For example:

In [19]:
3

3

In [20]:
print(3)

3


However, it is a function that is used ALL THE TIME in Python, mostly for debugging code. Like when you want to know what value is behind a certain variable. 

For now, it is enough to know that it is a function, and that it just "prints" what ever value it is given to the user :)

### 3.1 Assigning variables

Now to another fundamental componant of Python – _variables_. 

Variables can maybe best be described as names we give to values in our code. This is done with the **assignment operator** `=`:

In [8]:
number = 3

Above, we _assign_ the name `number` to a value – the integer 3. This is called an **assignment statement** in python.

The variable name `number` is now registered in the computers memory and points to the integer value 3. It is sort of like a nametag! 

So, if we type `number` in our code, the value 3 will be returned!

In [51]:
number

3

We can use our variables in expressions with operators:

In [52]:
number * 5

15

The name of a variable is saved to the computers memory, in what's called the variables **namespace**. Namespaces are kept seperated so that the computer doesn't mix up what's what in your code. 

This means that naming variables in Python is bound by certain rules. For example, names cannot be seperated by blank space:

In [53]:
number one = 3

SyntaxError: invalid syntax (<ipython-input-53-d7b288421347>, line 1)

...or include special characters:

In [148]:
number%one = 3

SyntaxError: cannot assign to operator (<ipython-input-148-911e113b032f>, line 1)

We can do operations using only variables:

In [9]:
number1 = 3
number2 = 5

number1 * number2

15

As a final point on simple assignment statements. In the Python community, _readability_ is seen as a virtue. Therefore: **always name your variables in a readable and logical manner.** This is to make the code more pleasent to read and easier to understand.

In [80]:
a = 3 # not a particularly good variable name
stuff = 3 # also not very helpful
num = 3 # better!
int_number = 3 # even better! (maybe a little long)

### 3.2 Variables and strings

Of course, we can also use variables when doing string concatination:

In [59]:
string1 = "Hello "
string2 = "world!"

string1 + string2

'Hello world!'

But there are many important ways we can use variables and strings together. So this section will go over string values a bit more thoroghly, starting with long strings:

#### 3.2.1 Long strings

Sometimes we want to assign very long string values to variables. To have one long string on one row is not very readable:

In [150]:
"Hello everyone! I am very happy that you are trying out this course. Also, this is a long string with no line breaks. Thanks!"

'Hello everyone! I am very happy that you are trying out this course. Also, this is a long string with no line breaks. Thanks!'

If you want to create a long string value, you can pass three quotation marks `"""` on each side of the text. Within those, you can do line breaks etc:

In [7]:
"""Hello everyone!

I am very happy that you are trying out this course.
Also, this is a long string with line breaks.

Thanks!

"""

'Hello everyone!\n\nI am very happy that you are trying out this course.\nAlso, this is a long string with line breaks.\n\nThanks!\n\n'

In the output above, you see the character `\n` – this is just the line break symbol. If we use the `print()` function, we will get the string value as we typed it:

In [9]:
print("""Hello everyone!

I am very happy that you are trying out this course.
Also, this is a long string with line breaks.

Thanks!
""")

Hello everyone!

I am very happy that you are trying out this course.
Also, this is a long string with line breaks.

Thanks!



#### 3.2.2 String formatting – inserting variables into string values

In Python, it is very common to insert variables in string values. Let's say we have a name that we would like to insert into a sentence:

In [151]:
name = "Johan"
sentence = " is a pretty nice guy!"

We could use string concatination to create a full sentence:

In [152]:
name + sentence

'Johan is a pretty nice guy!'

Well, thank you!

But there is an even cleaner way of doing it where we don't need an operator. We can use **string formatting**. There are various ways to do string formatting, but the latest is called **f-strings**.

To do this we need the key letter `f` just before the string quotations, like so: `f""`. This tells the interpreter that the string will be formatted with variables. To include variables, we insert them inside curly brackets `{}` within the string. Using string formatting, we can thus type: 

In [62]:
f"{name} is a pretty nice guy!"

'Johan is a pretty nice guy!'

..or just:

In [154]:
f"{name}{sentence}"

'Johan is a pretty nice guy!'

**CAUTION!** This is a very very useful technique. But the syntax (=the structure of statements in the code) of string formatting has varied in different versions of Python. This means the f-string syntax won't work in older versions of Python. The older syntax, however, is still acceptable in newer versions.

I won't go through the these in detail, just show you how the same string can be produced with string formatting in older versions of Python:

In [66]:
f"{name} is a pretty nice guy!" # f-string, available in Python 3.6 and later

'Johan is a pretty nice guy!'

In [67]:
"{} is a pretty nice guy!".format(name) # introduced in Python 3.0, now works in all versions 2.6 and later

'Johan is a pretty nice guy!'

In [68]:
"%s is a pretty nice guy!" % name # older than 2.6

'Johan is a pretty nice guy!'

I highly recommend the f-string, since it is way more readable, clean and easy to write.

#### 3.2.3 (Deepdive) – Long strings, string literals and the escape character `\`

##### 3.2.3.1 More on tripple quote strings

Above, we briefly touched the subject of creating long strings with tripple quotes `"""`. In these, we tell the interpreter that _all_ characters following these quotes are part of one string value, until it finds a the enclosing set of tripple quotes. This means we can do linebreaks within the string to make it more readable since we don't want the entire string on one line. Take this example:

In [156]:
# This isn't very readable...
single_quotes = "This is a long string. This is a long string. This is a long string. This is a long string. This is a long string. This is a long string. This is a long string. This is a long string. "

# ...compared to this:
tripple_quotes = """
This is a long string. 
This is a long string. 
This is a long string. 
This is a long string. 
This is a long string. 
This is a long string. 
This is a long string. 
This is a long string. 
"""

As you can see, the same string is a lot more readable using the `"""`. We should always try to make the code more readable (more on this in section 3.5), and as a general guidance: **never write lines of code that are more than 70 characters long**. The string value assigned to `single_quotes` is thus to long. The `tripple_quotes` string value is much more readable!


However, _they are not the same string._ The reason is because we did linebreaks in the second one. The difference becomes apparent if we print the two variables:

In [157]:
print(single_quotes)

This is a long string. This is a long string. This is a long string. This is a long string. This is a long string. This is a long string. This is a long string. This is a long string. 


In [158]:
print(tripple_quotes)


This is a long string. 
This is a long string. 
This is a long string. 
This is a long string. 
This is a long string. 
This is a long string. 
This is a long string. 
This is a long string. 



##### 3.2.3.2 String literals and the backslash `\`

If we want a more readable code using tripple quotes, but don't want to include any linebreak characters, we can use the backslash `\` character. This is a special character that lets us escape so called **string literals**. [There are a handful that you can use](https://docs.python.org/2.0/ref/strings.html), and they help us format strings in different ways. I won't go through them all here, but we've already covered one – the newline string literal `\n`: 



In [159]:
print("Hello\nworld!") # \n means "insert newline"

Hello
world!


Also, the horisontal tab `\t`:

In [160]:
print("Hello\tworld!")

Hello	world!


The newline string literal is actually inserted in tripple quote strings every time we press <kbd>Enter ↵</kbd> within our string quotes:

In [161]:
"""
"""

'\n'

So, back to the example we had in the variable `tripple_quotes`, we actually included the newline `\n` in the string on every row:

In [162]:
tripple_quotes = """
This is a long string. 
This is a long string. 
This is a long string. 
This is a long string. 
This is a long string. 
This is a long string. 
This is a long string. 
This is a long string. 
"""

In [163]:
tripple_quotes

'\nThis is a long string. \nThis is a long string. \nThis is a long string. \nThis is a long string. \nThis is a long string. \nThis is a long string. \nThis is a long string. \nThis is a long string. \n'

This means that if we were aiming to rewrite the variable `single_quotes` with `"""`, we failed, since it is a different string value:

In [164]:
single_quotes

'This is a long string. This is a long string. This is a long string. This is a long string. This is a long string. This is a long string. This is a long string. This is a long string. '

Using a comparison operator (more on these in the next chapter, they return True if values on each side of the `==` operator are the same), we can see that they are not the same value:

In [165]:
single_quotes == tripple_quotes

False

However, we can nullify string literals by using **the escape character**, the backslash `\`:

In [166]:
"""\
"""

''

Using the backslash, we can still use tripple quotes, and get the same value! It is handy while working with long string values – to increase readability while at the same time tell the interpreter which linebreaks that should include newline characters. In this case, we can escape all newlines to get `tripple_quotes` and `single_quotes` to produce the same exact string value.

So, if we want Using the `\` character:

In [167]:
tripple_quotes = """\
This is a long string. \
This is a long string. \
This is a long string. \
This is a long string. \
This is a long string. \
This is a long string. \
This is a long string. \
This is a long string. \
"""

In [168]:
print(tripple_quotes)

This is a long string. This is a long string. This is a long string. This is a long string. This is a long string. This is a long string. This is a long string. This is a long string. 


In [169]:
tripple_quotes == single_quotes # tada!

True

The mixture of tripple quotes and the `\` character is more for situations where you want one passage in the string to remain on one line when being printed, while some other parts of the same string should be printed on a new line. Take this example:

In [170]:
long_string = """
This is one long sentence, with a lot of words, \
that I still want to be kept in one long sentence; \
but since it is rather long I will need to cut it \
up to increase readability.

This, however, is a list within the string, \
and here I WANT linebreaks:

- First item
- Second item
- Third item
etc...

"""

In [171]:
print(long_string)


This is one long sentence, with a lot of words, that I still want to be kept in one long sentence; but since it is rather long I will need to cut it up to increase readability.

This, however, is a list within the string, and here I WANT linebreaks:

- First item
- Second item
- Third item
etc...




As a final point, the backslash is also called the **line continuation character**. If NOT used in a string value, the backslash tells the interpreter "this line of code continues on the next line"! Take this example where we concatinate a long string:

In [172]:
# This will fail
"Hello " +
"world! " +
"How " +
"are " +
"you?"

SyntaxError: invalid syntax (<ipython-input-172-e8beae0e0012>, line 2)

But if we use the line continuation character:

In [173]:
"Hello " + \
"world! " + \
"How " + \
"are " + \
"you?"

'Hello world! How are you?'

##### 3.2.3.3 Exercise – write a letter!

In this exercise, you will need to reformat the following text (it is a grocery list) so that it gets printed in  the correct manner. You will also need to heed of the rule to not have any line of code longer than ~70 characters. The parts enclosed within `<>` shall be replaced by whatever is instructed using variables and string formatting! 

Here's the letter with the grocery list:

In [44]:
grocery_list = "Hi <receiver>! All ingredients are necessary, \
so please don't forget any item in the list. Also, you know I love \
a little snack after dinner, so please get something you know I \
love – like <nice_treat>. Please hurry, I am starving! - Milk - \
Spinach - Strawberry jam - Spam - Whole cream - Cat food Remember \
that we have loads of credits to spend. We need to save up some \
money for our dream vacation next year. I am soooo looking forward \
to travel to <destination>. That is, if the virus ever would let us go :( Love, \
<your_name>"

In [45]:
print(grocery_list)

Hi <receiver>! All ingredients are necessary, so please don't forget any item in the list. Also, you know I love a little snack after dinner, so please get something you know I love – like <nice_treat>. Please hurry, I am starving! - Milk - Spinach - Strawberry jam - Spam - Whole cream - Cat food Remember that we have loads of credits to spend. We need to save up some money for our dream vacation next year. I am soooo looking forward to travel to <destination>. That is, if the virus ever would let us go :( Love, <your_name>


At the moment, it is on one single line. You will need to format it using these variables and string formatting:

In [46]:
receiver = ""
nice_treat = ""
destination = ""
your_name = ""

Here's how it should look:

In [48]:
# DON'T RUN THIS CELL


Hi Mom!

All ingredients are necessary, so please don't forget any item in the list. Also, you know I love a little snack after dinner, so please get something you know I love – like Ben & Jerrie's ice cream. Please hurry, I am starving!

- Milk
- Spinach
- Strawberry jam
- Spam
- Whole cream
- Cat food

Remember that we have loads of credits to spend. We need to save up some money for our dream vacation next year. I am soooo looking forward to travel to Paris. That is, if the virus ever would let us go :(

Love,

Johan



Remember that you can use string formatting, tripple quotes strings and the escape backslash `\`! The solution can be found in the solutions notebook!

### 3.3 Changing variables – Operators and assignment statements

Often when coding, we want to change the values of our variables. We _overwrite_ our variable, giving it a new value. This is easily done in python:

In [67]:
number = 3 # assignment statement

In [68]:
number = number + 4

In [69]:
number # "number" is now 7 (3 + 4 = 7)

7

#### 3.3.1 Augmented assignment operators

When using the assigment operator `=`, the python interpreter will evaluate the expression to the right of the assignment operator first. Then it will assign the returned value to the namespace written to the left of the `=` (the variable name `number`). This is how you reassign, or overwrite, your variables. 

But you can also use the assignment, AND the addition operator in one go using `+=`. This is called an **augmented assigment operator**, it uses the operator on the variable _while also assigning the result to the variable_. Take this example:

In [174]:
number = 3

In [175]:
number += 4 # '+=' is syntax for "add the value to the right to the variable name to the left"

In [176]:
number

7

The same can be done with subtraction, multiplication and division:

In [177]:
number -= 1

In [178]:
number

6

In [179]:
number *= 5

In [180]:
number

30

In [181]:
number /= 3

In [182]:
number

10.0

### 3.4 Multiple assignment statements

In Python, let's say we want to create four variables with different values. This is boring and tedious to type out:

In [10]:
one = 42
two = "Johan"
three = True
four = "elephant"

In [11]:
one

42

In [12]:
four

'elephant'

Luckilly, this can be written in one single line of code, with a single `=`. We just pass the variable names to the left of the assignment operator – seperated by commas – and the values to the right – seperated by commas – in the order we want the values assigned:

In [13]:
one, two, three, four = 42, "Johan", True, "elephant"

In [14]:
one

42

In [15]:
two

'Johan'

In [16]:
three

True

In [17]:
four

'elephant'

### 3.5 (Optional) More about readability – the zen of Python

Earlier, we talked about naming variables in a logical and transparent manner. We should name our variables making it easier to understand what values they may hold. This is important – it makes your code more readable. Readability is a very important aspect of writing Python. 

The reason for this is quite simple: **as a person working with coding, you will read a lot of code.** Returning to old projects for example, what code does what and where in your program? If you learn early on to make your code more readable, you will save eons of time later on!

The Python community will sometimes refer to what is called [_the zen of python_](https://www.python.org/dev/peps/pep-0020/). This is the general "rules" that should guide you when you code. These are the first seven lines in it:

```
Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
```

This is a bit abstract, but it is words that is good to keep in mind going forward. There are many concrete ways to write more readable code. For example, use white space (blank space) in your expressions:

In [86]:
2 + 2 # zen-like

4

In [87]:
2+2 # this is less readable – therefore, don't! :)

4

Or, in your assignment statement:

In [88]:
num=2 # don't
num = 2 # do!

For more on how to write readable python code, [check out this link!](https://docs.python-guide.org/writing/style/)

### 3.6 Exercise – string concatination and variables

In this exercise, I want you to assign these seven string values to one variable each:
- `"was"`
- `"the"`
- `"on"`
- `"cat"`
- `"mat"`
- `"inside"`
- `"hat"`

Using these variables, use string concatination to produce these three sentences:
1. "cat was inside the hat on the mat"
2. "hat was inside the cat on the mat"
3. "mat was inside the hat on the cat"

You can add blank spaces to the strings if you'd like.

When assigning your variables, try to use a multiple assignment statement.

The solutions is found in the solutions notebook (there is more than one!

---

## 4 Comparing values

### 4.1 Comparison operators

In section 1, we briefly covered the boolean values. In code, we often want to check if something is true or false. 

Let's say we want to write a program that functions as an alarm that lets us know if our workday is over. The program would then have to check the time, and then see if the present time is past five o'clock. If so, it alarms us that it's time to head home. If not, it does nothing. Lets put this down in writing:

Program workflow:
```
1. Check time
2. See if time is past 17.00 (either true/false)
3. If true -> alarm us! If false -> do nothing
4. Terminate program
```
In such a program, we need an operator that can compare values. In our case, compare present time with the end of our workday (17.00). In Python, we can use a **comparison operator** to do this:

```
COMPARISON OPERATORS

    ==   Equal to
    !=   Not equal to
    <    Less than
    >    Greater than
    <=   Less than or equal to
    >=   Greater than or equal to
```
Expressions with comparison operators always return a boolean value:

In [99]:
# 3 is equal to 3
3 == 3 

True

In [100]:
# 3 is not equal to 4
3 == 4 

False

In [103]:
# 3 is less than or equal to 3
3 <= 3 

True

In [101]:
# 3 is less than or equal to 4
3 <= 4 

True

In [18]:
# 3 is not greater than or equal to 4
3 >= 4 

False

Note that comparison operators _also checks for value types_. This means that:

In [108]:
3 == "3"

False

...since `3` is an integer and `"3"` is a string. However, integers and floats can be equal to eachother:

In [109]:
3.0 == 3

True

Comparison operators can, of course, also be used to check variables:

In [105]:
my_name = "Johan"

In [106]:
my_name == 'Pete'

False

Above, please note that the first cell `my_name = "Johan"` has _one_ assignment operator. This is an assignment statement where we create our variable. The second cell has _two_ assignment operators – hence, it is a comparison operator.

### 4.2 Boolean operators

There are also three boolean operators: `and`, `or` and `not`. These let us evaluate two or more booleans into one boolean value. Since we often need to take more things into consideration than something _only_ being true or false, these operators are essential when coding. Programs and scripts often need to handle ambigious situations.

#### 4.2.1 The `and` operator

Consider our earlier example, our workday alarm program. Let's say the workday starts at 9.00 in the morning and we want to check if the present time is during work hours or not. If so, we can use a boolean operator: is the present time after 09.00 _AND_ before 17.00?

Let's first look at the `and` operator. It returns `False` if the two expressions to its left _and_ to its right  isn't true:

In [186]:
1 == 1 and 4 == 4

True

In [127]:
1 == 2 and 4 == 4

False

In [128]:
1 == 2 and 4 == 5

False

In [183]:
True and False

False

In [184]:
True and True

True

In [185]:
False and False

False

In other words, the `and` operator will always evaluate to `False` if both of its expressions aren't `True`.

#### 4.2.2 The `or` operator

If we instead use the `or` operator, it will return `True` if either expressions are true:

In [122]:
1 == 2 or 4 == 4

True

In [120]:
True or False

True

In [124]:
1 == 2 or 4 == 5

False

In [123]:
False or False

False

#### 4.2.3 The `not` operator

The `not` operator only evaluates on one boolean value, and it always returns its opposite:

In [129]:
not True

False

In [130]:
not False

True

#### 4.2.4 Boolean operators – examples

This may look like very abstract concepts, but it makes way more sense if we use them together with variables. Let's create some:

In [189]:
name = "Johan"
nationality = "Swedish"

Using it together with the `and` operator:

In [190]:
# is 'name' equal to "Johan", and 'nationality' equal to "Swedish"?
name == "Johan" and nationality == "Swedish"

True

In [191]:
# is 'name' equal to "Johan", and 'nationality' equal to "Finnish"?
name == "Johan" and nationality == "Finnish"

False

The `or` operator:

In [192]:
# is 'name' equal to "Johan", or 'nationality' equal to "American"?
name == "Johan" or nationality == "American"

True

`name` is equal to "Johan", therefore the expression evaluates to `True`.

The `not` operator:

In [137]:
# Is 'name' not equal to "Pete"?
not name == "Pete"

True

Let's use these concepts to help us with our alarm program:

In [138]:
# Let's say the time is presently 12.15
presentTime = 12.15

In [139]:
# Is the present time before 17.00, but after 9.00?
presentTime < 17.00 and presentTime > 9.00

True

In [140]:
# Let's say the time is presently 17.15
presentTime = 17.15

In [141]:
presentTime < 17.00 and presentTime > 9.00

False

### 4.3 Exercise – insert the correct operators

In this exercise, I want you to uncomment the code. Then, replace the `?` characters with the correct operators so that the `print()` function is executed. In other words, you'll need to change operators so that all comparisons are evaluated to `True`. 

These code cells include `if`-statements, which will be the subject in the next chapter!

In [1]:
number = 12

In [ ]:
#if number ? 15:
#    print("Correct!")

In [2]:
#if number ? 5 and number ? -5:
#    print("Correct!")

In [ ]:
#if number ? 150 or number ? 25:
#    print("Correct!")

In [3]:
#if number ? 12:
#    print("Correct!")

In [71]:
#if number ? "Hello world!":
#    print("Correct!")

## 5 Flow Control Statements

As has been discussed above, we often want to take context into consideration when coding: "If this expression is true, do this thing...". This often means we want to have control of our program's progress, how it "flows". 

In Python, we can use comparison operators in **flow control statements**: these are: 
- `if` 
- `elif`
- `else`

\*\* Going forward, remember that **what you type (the syntax) is important!** Be sure that you write these statements exactly as described. Otherwise, you will get an error and your code won't work. \*\*

### 5.1 The `if`-statement

Flow control statements lets you decide a certain logic for your code to follow. The `if`-statement consists of the keyword `if`, followed by a **condition** (=an expression in a flow control statement that evaluates into `True` or `False`), and ended with a colon `:`. 

If the if-statement's condition evaluates to false, the interpreter skips the if-statement's code. Have a look at this:

In [222]:
name = "Johan"

In [223]:
if name != "Johan":
    print("Hi Johan!")

Code indented at the same level is called **a block of code**. The expression `name != "Johan"` is evaluated to false, so above if-statement's isn't executed. It is basically the same as typing:

In [224]:
if False:
    print("Hi Johan!")

...since `name != "Johan"` evaluates to False.

Indentations are very important in Python. It tells our computer which code is grouped together. All code that is indented directly under an if-statement is that if-statement's code block! The `print()` function is one block of code within our if-statement (even though, in this example, it's just one line).

If the condition in the if-statement is evaluated to `False`, the indented block (containing `print("Hi Johan!")`) will be skipped and not be executed. If the condition is `True`, then the indented code will execute:

In [226]:
if name == "Johan":
    # The condition name == "Johan" is true! Hence, this will be executed:
    print("Hi Johan!") 

Hi Johan!


Another example:

In [227]:
age = 35

In [202]:
if age > 30:
    # True! This will be executed
    print("Whoa! Not that young any longer, buddy!") 

Whoa! Not that young any longer, buddy!


In [203]:
if age < 30:
    # This will be skipped since 35 is more than 30
    print("Whoa! Not that young any longer, buddy!") 

In [204]:
youngAge = 25

In [205]:
if youngAge < 30:
    print("You, sir, are in your prime!")

You, sir, are in your prime!


We can also use boolean operators in flow control statements, like the `and` operator:

In [206]:
if youngAge < 30 and youngAge >= 20:
    print("You are in your twenties! Congratulations!")

You are in your twenties! Congratulations!


### 5.2 The `else`-statement

But if we also want to cover situations where the if-statement is false? Then we use the `else`-statement:

In [207]:
age = 35

In [208]:
if age < 30:
    # Since age < 30 is false, this code will be skipped
    print("Hello young sir!")
else:
    # Instead, this will be executed
    print("Oldie but goldie!")

Oldie but goldie!


So "If an expression is true, do something... Otherwise, do this other thing..." An `else`-statement consists of the `else` keyword followed by a colon, `:`. SInce it covers ALL situations not specified by the if-statement, no expression is needed.

An else-statment does only work if it follows an if-statement. If you write an else-statment by itself, it will produce an error: 

In [172]:
else:
    print("Hello world!")

SyntaxError: invalid syntax (<ipython-input-172-eb2670b32139>, line 1)

Our workday alarm program, but with an if/else-statement:

In [211]:
present_time = 12.15

In [213]:
if present_time < 17.00 and present_time > 9.00:
    print("You should be working!")
else:
    print("Not working hours!")

You should be working!


### 5.3 The `elif`-statement

In many situations we want to explicitly cover more scenarios than just those in our if- and else-statements. We can use `elif`-statements:

In [214]:
age = 16

In [215]:
if age >= 30:
    # 16 is lower than 30, so this will be skipped
    print("You are not very young any longer, I'm afraid.") 
    
elif age >= 20 and age < 30:
    # 'age' is not between 20 and 29, this block is also skipped
    print("You are in your twenties!") 
    
elif age < 20 and age >= 13:
    # Bingo! This code will be executed
    print("A teenager!") 
    
else:
    # This would execute in all situations where 'age' is less than 13
    print("You are very, very young!")

A teenager!


As you can see in the code cell above, we can use more than one elif-statement. You can put in as many as you like! 

Just like the else-statement, elif-statements can only be used after an if-statement. Otherwise, an error will be produced:

In [196]:
elif age >= 20 and age < 30:
    print("You are in your twenties!")

SyntaxError: invalid syntax (<ipython-input-196-5150a1770b0d>, line 1)

Here is the hierarchy of these three flow control statement:

1. if-statement goes first
2. elif-statements (as many as you like!) comes second
3. And finally, the else-statement

### 5.4 (Optional) "Truthy" and "Falsy" values

Like covered above if-statements need a condition (=expression) to evaluate to true or false to run. But in Python, all values are actually either truthy or falsy. This means that a single value following the if-keyword will also work:

In [216]:
if 3:
    print("Works!")

Works!


In [217]:
if 3.5:
    print("Works!")

Works!


In [218]:
if 0:
    print("This won't be printed...")

In [219]:
if 0.0:
    print("This won't be printed...")

In Python, all number values, that is both integers and floats, above 0 is considered "truthy" values. 0 and 0.0 is "falsy". In string values, all empty strings `""` are considered falsy, all other string values are truthy:

In [220]:
if "a string value":
    print("Works!")

Works!


In [221]:
if "":
    print("This won't be printed...")

Considering we should always strive to make our code more readable, above examples should be avoided. Instead we should write an actual comparison:

In [207]:
value = "string"
if value != "":
    print("This wasn't an empty string!")

This wasn't an empty string!


We can always check values "truthiness" with the built-in function `bool()`:

In [213]:
bool(3)

True

In [214]:
bool("")

False

In [215]:
bool(0)

False

In [216]:
bool("Hello world!")

True

### 5.5 Exercise – type the end of the story!

If-statements are rather useful if you're creating small text based games. Think of a Python script that runs a dungeons and dragons game, what does the player choose? How does it affekt the storyline?

In this exercise, you will write an ending to a little (silly) story. You will write it as an flow control statement. By choosing a value for the variable `item`, and by passing it to a flow control statement, this variable decides the ending of the story! Depending of the value of `item` a different ending should then be printed with a `print()` function. You write the story ending that's printed! Save this story to variable called `ending`, and then print this variable. Simple!

Remember that you can write string values with tripple quotes `"""` to make them a long string (see section 3.2 if you want to refreash your memory!). Good luck!

**The End of the Story**

```
Teddy Merrybottom was tired. He had done what had been asked of him. With him, he carried enough booty to make his entire village the most wealthy in Southern Illyria. But the journey to Doom Mountain had been exhausting, and tragic. His three best friends had all perished. 

He couldn't for the life of him understand why they wanted to swim while in full armor. Granted, their fight with the dragon Grymf had been a fantastic, and surprising success – worthy of great celebration. But getting completely planked on honey beer, and then go swimming with breast plates? It was a bummer. Well, it was also sad, since, you know, they died.

Even more disapointing was that the only one in Teddy's company that hadn't joined the party – and died – was Gary. And Garry was an idiot. And now, even though their three best friends was dead, Gary just kept complaining about his arthritis and how his knees hurt. This also meant that Teddy was taking care of all their stuff.

Fortunatly, they were just about a hill away from their village. Teddy looked forward to being in front of his fire place, being rich and all that. But most of all, not having to spend any more time with Gary.

– What is that!? Gary suddenly gasped

– Where? Teddy replied, reaching for his sword hilt

They were walking on a field right next to the edge of Hillside Forrest, and inbetween the trees, Teddy saw the contours of something vile. It was a Vampwolfosauros. A howling, bloodsucking lizard beast. It had seen them, and rushed towards them in a rage, trampling trees in its path.

Teddy, struck by an instant adrenaline rush, drew his sword. He turned around to get his shield from the horse carriage. But in his exitement, he stumbled, and grabbed the only thing close to help him not loosing his footing – the horse strings. 

This pulled the horse in the direction towards the approaching monster, and, of course, it panicked and reared. It did so right next to a small boulder with a sharp edge, which cut the horse's leg. The horse, therefore, also lost its footing, tumbling over Teddy. In the chaos, the cart crashed to its side. 

Suddenly, Teddy found himself stuck in a pile of gold, jewels, weapons, and horse. And the Vampwolfosauros was just moments away. It was sure to kill them all! 

It was now all in the hands of Gary. Teddy and Gary both realised this, and they both felt terror. Gary's only chance was that Teddy would throw him something for their mutual protection against the massive, raging beast. But what?

Teddy looked around. He saw four magical items lying next to him: 
- the Ice Jewel
- the Ancient Staff of Pharts 
- the book "The World's 10 Worst Poems"
- the Bear Spear (a magical spear shaped like a bear)

He realised that chances were slim with either, since Gary was an blidgering moron. But he reached out and grabbed his item of choice and threw it to Gary, screaming "Come on you fool, save us!!"
```

In [ ]:
item = "" # <- which item does Teddy choose?

In [11]:
# write a flow control statement, containing at least one each of an if-, elif- and else-statement
# this flow control statement should decide the end of the story, depending on which item Gary's given
# Good luck!

## 6 Functions

### 6.1 Definitions and terminology

A **function** in Python is perhaps best described as a miniprogram within our code. It is block of code that is assigned to a namespace. We can then use and reuse the function, without having to rewrite all the code within it! Above, we have briefly touched on the subject of Pythons built-in functions, and in this chapter we will go through more of them. But we will also learn how to write our own functions.

But to start, why do we need functions? The short answer is that they save us alot of time and effort when writing programs.

As an example, let's say we want to count the number of characters in a string value. We can write code to do that for us! 

_(In this example, don't get stuck on the code itself, it includes concepts we still havn't covered)_

In [229]:
string_value = "Johan"

In [233]:
# again, don't get stuck in this code. We will cover it later :)
for index,char in enumerate(string_value):
    num = index + 1 

print(f"The string is {str(num)} characters long")

The string is 5 characters long


Above, we have a small snippet of code that helps us calculate how many characters there are in a string value. But let's say we want to use this code snippet in many different places in our program. It would be very tiresome to write it over and over again. Instead, we could use a function instead! Fortunately, there is a built-in function that helps us with excactly what we're after – the `len()` function:

In [275]:
len("Johan")

5

The `len()` function returns the number of items in a python container object. So, if given a string, the `len()` function will return the number of characters in the string. As you can see, the use of `len()` saved us alot of code typing!

A function always have a name. The `len()` function has the name "len". Functions can be used by typing parenthesis `()` directly after the function name. In Python, when using a function, we say that we **call the function** – or we do a **function call**. But we can also say that we **invoke the function**. As you can see, there are often many names to the same concepts in Python. 

The string value "Johan" we wrote within the parenthesis is called **an argument**, and when we give functions arguments, we say that we **pass arguments to functions**. 


(These phrases will be used throughout this notebook, but if you are uncertain of the meaning of a term used, first try finding the explanation wihtin this notebook using <kbd>⌃ Control</kbd>+<kbd>F</kbd>, OR Google it! Googling programming concepts is a great way to learn!)

In Jupyter, we can type <kbd>⇧ Shift</kbd>+<kbd>⇥ Tab</kbd> when our marker is within a function's parenthesis to see information about the function!

In [281]:
len("") # <- try putting your marker within the double quotes in the parenthesis and tapping shift+tab

(The information you see when tapping <kbd>⇧ Shift</kbd>+<kbd>⇥ Tab</kbd> is called the function's **docstring**)

### 6.2 Built-in Functions

In previous chapters in this notebook, we have used some built-in functions already. For example:

In [282]:
type(42) # checking an objects datatype

int

In [226]:
print("Hello world!") # tells the interpretor to print the function's argument

Hello world!


We can also convert values into other datatypes (remember that, in a notebook, you can always check the function's docstring pressing <kbd>⇧ Shift</kbd>+<kbd>⇥ Tab</kbd> when your marker is within the functions' parenthesis):

In [227]:
str(4) # returns passed argument as a string

'4'

In [228]:
int(5.0) # returns passed argument as an integer

5

In [230]:
int("5")

5

In [233]:
float(10) # returns passed argument as a float

10.0

In [234]:
float("10")

10.0

There are many more of these built-in functions. Every time you use python code, these will work from the getgo. But there are many many more functions that come prepackaged within the standard Python code package, but that don't work immediatly. They have to be imported from various modules.

### 6.3 The standard library and modules

When you use a freshly installed version of Python, it comes with a set of built-in functions. But it also comes with a variety of "libraries" of code that can be easly accessed and used in your code. This is called **the standard library**. To get access to these, we use the `import` statement. The libraries of code we import are called **modules**, or **packages**. Let's import the `random` module:

In [236]:
import random

After we have imported, we can use functions in this library by typing a dot `.` directly after the module name (using dots to access methods is called **dot notation**). 

In Jupyter, there is a really neat trick to see which functions are included. After having typed `random.`, press the <kbd>⇥ Tab</kbd> key to get a list of available functions:

In [237]:
# replace this comment with the code "random." and then place your marker next to the "." and press the tab key!

In this case, we want to use the function called `randint()` from the random module. This function takes two arguments: one starting point number value, and one end point number value, and returns a random number inbetween:

In [238]:
random.randint(1, 10)

2

In [239]:
random.randint(-2446, 10645)

5864

That is how you use imported modules and their functions!

We can also specifiy which method we want from the random module in our import statement by using the `from` keyword:

In [78]:
from random import randint # from the 'random' module/package, import the function 'randint'

In [79]:
randint(1, 10) # now, we don't need to specify module, since we directly imported the function itself!

3

We can also use aliases of packages. This is done with the `as` keyword. This is usually done to save typing time by shortning long names. For example:

In [240]:
import datetime as dt

Now we only type `dt` to get access to the package's methods and classes:

In [243]:
dt.datetime.today()

datetime.datetime(2020, 10, 4, 21, 25, 54, 694958)

Another cool module to import is the `os` module. It lets us explore the files on our computer! Let's import it:

In [80]:
import os

We can use the `listdir()` function to get us a list of all files in the directory (=folder) this notebook is located within:

In [81]:
os.listdir()

['python_intro_eng.ipynb',
 '.DS_Store',
 'LICENSE',
 'course_material',
 'Pipfile',
 'solutions.ipynb',
 '.gitignore',
 '.ipynb_checkpoints',
 '.git',
 'Pipfile.lock']

In our import statement, we can also just import all functions and methods within a module at the same time with the from keyword. This is done using the `*`

In [244]:
from os import *

This means we imported ALL functions within the `os` module. It is the same as just typing `import os`, but now we don't need to specify the module name before using its functions. 

HOWEVER! This is not recommended. It makes our code less readable. In Python, you should always aim to make your code transparent.

#### 6.3.1 Exercise – the coin toss

The random module's function `randint` can be used to produce random number. In this exercise, you shall try to build a small script that takes a random number, and depending on the number, either "Heads!" or "Tails!" is printed by the `print()` function. This will require you to use variables, the `random.randint()` function, and flow control statements.

Remember to import the random module! The solution is found in the solutions notebook.

Good luck!

### 6.4 Third party modules

One huge advantage in learning Python is its flourishing community. Many, many, many programmers contribute to the Python language, often by supplying third party modules. They are packages that is not part of the standard library. They need to be downloaded before you can import and use their functions. This is normally done by using the `pip` Python library installer, or through Anaconda. However, this is not covered in this notebook, but you can find an extensive guide [here.](https://packaging.python.org/tutorials/installing-packages/)

If you are using Jupyter through the Anaconda software (as is recommended), this means you acctually have a number of such third party modules already installed on your computer. You can see all your available libraries by using the built-in function `help()`:

In [ ]:
help('modules')

In the list you can find libraries such as `pandas`, `numpy` and `matplotlib` (to name a few) – all third party modules! All the packages in the list can be imported using the `import` statement.

### 6.5 Writing your own functions

As was explained briefly in the beginning of this chapter, a function is like a miniprogram within our code. This miniprogram has a name that we can call (or invoke) to use and reuse the code block within the function.

In the above sections, we have learned about built-in and imported functions. But we can build our own! Let's build a function that greets us hello:

In [244]:
def greetings():
    print("Hello deer!")
    print("How are you?")
    print("I am well, thanks for asking!")

Running above cell will not produce an output. It is a function statement where we simply define our function. This is done by typing the keyword `def` (for "define"), followed by our name of the function. Then enclosed parenthesis, followed by a colon `:`. 

When you've typed all that, it is time to define the functions content code. After the colon, we press the <kbd>↵  Enter</kbd> key, and the next line is automatically indented, marking that the functions code is in a new code block! The function's content is the `print()` functions that are then called one after another. 

We can now call our function by invoking its name!

In [296]:
greetings()
greetings()
greetings()

Hello deer!
How are you?
I am well, thanks for asking!
Hello deer!
How are you?
I am well, thanks for asking!
Hello deer!
How are you?
I am well, thanks for asking!


Now, if we HADN'T used a function, the code for getting us the same output would be:

In [290]:
print("Hello deer!")
print("How are you?")
print("I am well, thanks for asking!")
print("Hello deer!")
print("How are you?")
print("I am well, thanks for asking!")
print("Hello deer!")
print("How are you?")
print("I am well, thanks for asking!")

Hello deer!
How are you?
I am well, thanks for asking!
Hello deer!
How are you?
I am well, thanks for asking!
Hello deer!
How are you?
I am well, thanks for asking!


In Python, we want to have as little duplicated code as possible. This means we get shorter programs, code that is easier to read, and – perhaps most importantly – _code that is easier to edit and debug_. 

As some of you might've spotted, there is a spelling error in above code. We've acctually not used the pleasantry "Hello dear" in our greeting. Instead we called the person we're greeting a deer. Not very nice. When using functions, this is an easy fix: we just change the code within the function's code block to:

In [291]:
def greetings():
    print("Hello dear!") # <- easy fix!
    print("How are you?")
    print("I am well, thanks for asking!")

In [293]:
greetings()
greetings()
greetings()

Hello dear!
How are you?
I am well, thanks for asking!
Hello dear!
How are you?
I am well, thanks for asking!
Hello dear!
How are you?
I am well, thanks for asking!


If we hadn't used a function, we would've had to change it in all places where the spelling error occured:

In [292]:
print("Hello dear!") # Here...
print("How are you?")
print("I am well, thanks for asking!")
print("Hello dear!") # here...
print("How are you?")
print("I am well, thanks for asking!")
print("Hello dear!") # and here
print("How are you?")
print("I am well, thanks for asking!")

Hello dear!
How are you?
I am well, thanks for asking!
Hello dear!
How are you?
I am well, thanks for asking!
Hello dear!
How are you?
I am well, thanks for asking!


As you can see, this is not very practical. And it is also easy to miss all the occurances of an error in duplicated code. So: if you can avoid it – don't duplicate code! :)

### 6.6 Arguments

In [297]:
len("Johan")

5

In this `len()` function call, we pass the function an argument – the string value "Johan". When building our own functions, we can specify if they require arguments. We do so by writing a variable name within the parenthesis while defining our function. This variable is called a **parameter**. Let's modify our greeting function:

In [353]:
def greetings(parameter):
    print("Hello dear " + parameter + "!") # this is string concatination. Remember?
    print("How are you?")
    print("I am well, thanks for asking!")

Arguments are information that is passed in the function call and is then used within the function's code block. Above, I've called the parameter `parameter` to make it obvious where we put it. We can, however, name the arguments to whatever we like. We want the code to be readable, and the name `parameter` is rather generic and not very readable. Let's change it:

In [245]:
def greetings(name):
    print("Hello dear " + name + "!")
    print("How are you?")
    print("I am well, thanks for asking!")

Now, when we call the function like before, we will get an error:

In [246]:
greetings()

TypeError: greetings() missing 1 required positional argument: 'name'

When a argument is mandatory for the function call to work, it is called a **positional argument**. The error message tells us that an argument is required, and then tell us the name of the parameter missing a value. This is why it is important to name your parameters in a readable manner. If our name would just be "parameter", the error message wouldn't make any sense.

When an argument is put in the function call, this will then be used in the function's code block. To be precise: the argument will be assigned to the function's parameter. In this case, the string value "Johan" (the argument), will be assigned to the function's parameter `name`:

In [247]:
greetings("Johan")

Hello dear Johan!
How are you?
I am well, thanks for asking!


We can now use the function with any name we'd like:

In [357]:
greetings("Pete")

Hello dear Pete!
How are you?
I am well, thanks for asking!


In [358]:
greetings("Hanna")

Hello dear Hanna!
How are you?
I am well, thanks for asking!


As you can see, we can pass any string we like to the function!

In [359]:
greetings("KJ&€!AEFHO&#AHIUFQO%!PJ")

Hello dear KJ&€!AEFHO&#AHIUFQO%!PJ!
How are you?
I am well, thanks for asking!


It is worth explaining once more: When we pass our string value to the function, it is assigned _within the function_ as a variable – the variable `name`. It is this variable that is called the function's parameter. We can even specify which parameter the argument should be assigned to in the function call:

In [350]:
greetings(name="Johan") # the parameter is set to the argument value "Johan"

Hello dear Johan!
How are you?
I am well, thanks for asking!


Here's another function that just prints the function's argument:

In [250]:
def return_name(name):
    print(name)

In [249]:
return_name("Johan")

Johan


So, to recap:
- The function's **argument** is the value that is passed to the function when you call it (the string value "Johan" in the above example)
- The function's **parameter** is the variable name to which the argument value is assigned (the variable `name` in the above example)

You can also have more than one parameter when you define your functions. You just put more parameters within the parenthesis when defining your function. Each parameter is seperated by a comma.

As an example, let's redefine our greeting function to also require a `country` parameter:

In [251]:
def greetings(name, country):
    print("Hello " + name + "! You're from " + country + ", right?")

Let's clean it up using a f-string (see section 3.2.2 if you need a refreasher):

In [252]:
def greetings(name, country):
    print(f"Hello  {name}! You're from {country}, right?")

In [253]:
greetings("Johan", "Sweden")

Hello  Johan! You're from Sweden, right?


Since we have defined the function with two parameters, just giving the function one argument in the function call will now produce an error:

In [336]:
greetings("Johan")

TypeError: greetings() missing 1 required positional argument: 'country'

Remember, the order of parameters counts. If you pass positional arguments in the wrong order, it may not result in what you expected:

In [337]:
greetings("Sweden","Johan")

Hello Sweden! You're from Johan, right?


However, if we specify which argument we want to which parameter variable name, the order doesn't matter:

In [338]:
greetings(country="Sweden", name="Johan")

Hello Johan! You're from Sweden, right?


#### 6.6.1 Exercise – truth-teller function

In this exercise, you will define your own function. This function shall have two parameters. When you call the function, it should first print the two arguments, then it should compare them. If the passed arguments are not the same, the function should print "The arguments are not the same...". If they are the same, the function should print "The arguments are the same!"

The solution can be found in the solutions notebook. Good luck!

### 6.7 (Optional) Styling conventions writing functions

Here's a (kind of) quick style guide when you're defining functions. Let's take the earlier example:

In [363]:
def greetings(name, country):
    print(f"Hello  {name}! You're from {country}, right?")

In [364]:
greetings(name="Johan", country="Sweden")

Hello Johan! You're from Sweden, right?


When specifying argument to parameters as in the cell directly above, note that we don't use blank space inbetween the `=`. 

Some of you might have noticed that this is different than when we defined variables in section 3 above. If we would simply define a variable like this:

In [339]:
name = "Johan"

...we use one blank space on each side of the `=` operator. It is common practice to _not_ use blank space when specifying parameters in function calls:

In [340]:
greetings(name="Johan", country="Sweden") # No blank spaces next to the "=" operators!

Hello Johan! You're from Sweden, right?


However, there IS blank space after the comma. All this is common practice in the Python community, since it is considered to increase your code's readability. It is would still work with blank space:

In [365]:
greetings(country = "Sweden", name = "Johan") # don't write like this, even though it works :)

Hello Johan! You're from Sweden, right?


Why? This is in accordance with [the official Python style guide.](https://www.python.org/dev/peps/pep-0008/#whitespace-in-expressions-and-statements)

Lastly, since the function `greetings()` _requires_ two argument – that is, two **positional arguments** – we shouldn't name them in our function calls. It is seen as overly wordy and verbose. So this is prefered:

In [366]:
greetings("Johan", "Sweden")

Hello Johan! You're from Sweden, right?


### 6.8 The `return` statement

Just having functions printing information may not be sufficient when you write a function. Often, you want your functions to do something to the values you pass to them – you want them to _return_ a result of some kind. We can specify what to be returned from functions we write using the `return` statement. As an example, we will build a small calculator!

In [254]:
def multiplies_by_three(number):
    number = number * 3
    return number

multiplies_by_three(5)

Let's go through this step by step:
1. We have defined a new function that we named `multiplies_by_three`
2. We've specified that the function has a parameter – `number`
3. The functions code block then multiplies the `number` parameter by three, and reassigns the results to the variable `number`
4. It then _returns_ the resulting number value that is assigned to `number` using the `return` keyword. _This is the return statement_
5. Finally, we call our function, passing it the integer value 5 as an argument. It runs the integer through the function's code block and returns the value 15

Since the function now returns a value, we can save this result in a new variable:

In [256]:
results = multiplies_by_three(5)

In [257]:
results

15

Actually, a function _always_ return a value. But if you don't specify which one with a return statement, it will return the `None` value. See what happens if we only remove the result statement:

In [379]:
def multiplies_by_three(number):
    number = number * 3

In [380]:
results = multiplies_by_three(5)

In [382]:
results # Doesn't return anything?

In [384]:
type(results)

NoneType

In [385]:
None == results

True

### 6.9 Exercise – the currency converter

In this exercise, we shall build a function that takes two arguments. The first is a number value specifying an amount of euros, and the second is a string value specifying which currency we want to convert into (preferably the currency codes). 

So, if we pass our function the arguments `2000` and `"USD"`, the function shall return the amount 2000 euros converted into American dollars. The function should be able to calculate at least four different currencies:
- Swedish krona (SEK), 1 euro equals 10.43 Swedish krona
- American dollar (USD), 1 euro equals 1.17 American dollars
- British Sterling Pound (GBP), 1 euro equals 0.92 British pounds
- Chinese Renminbi (RMB), 1 euro equals 7.89 renminbis

Remember to try to write the function as zen-like (also called "pythonic") as possible! For example, name your parameters in a logical manner :) 

The solution is found in the solutions notebook. Good luck!

### 6.10 Bulletpoints

- Functions are small miniprograms. They are containers of block of code. They help us not having to dublicate code
- We **call** functions by **invoking** them, which means we write their names followed by enclosed parenthesis
- Often, we can pass values to function when invoking them. These input values are called **arguments**
- The arguments' values are assigned to the function's **parameters**, which are then used as variables within the function's code block
- By invoking functions, it **returns a value as an output**. We can specify what that should be in the function by using **the return statement**

## 7 Local and global scopes

Hopefully, you now understand how to use and how to build your own functions, and why they are useful when coding in Python. But there is one concept we havn't introduced yet which is important to understand when writing functions in Python: **scopes**.

The code that is executed outside a function is in a different scope than the code living inside it. Take this function that creates a variable named `numbe`, assigns the integer 9 to this variable, and then returns it:

In [258]:
def make_nine():
    number = 9
    return number

Let's try it!

In [432]:
make_nine()

9

When we called the function, the code block within it was executed, and the integer 9 was returned. But what happens if we first create a variable outside the function, assigned to another variable with the same name:

In [259]:
number = 3

In [260]:
number

3

...and now run the function again:

In [261]:
make_nine()

9

So, if we now check the variable `number`, which value will we get? Can you guess?

In [262]:
number

3

_Why is this?_ 

We first created the variable `number` outside the function. This happend in the scope "furthest outside" – the ground level sort to speak – and this is called **the global scope**. There is no scope outside the global scope. Variables that are created in the global scope are called **global variables**.

The code that lives within the function is in a **local scope**. When the function is invoked, the code inside the function is executed in _a local execution context_. This means that the variable `number` we created inside the function is within in a different context than when we defined the global variable `number = 3`, in the global scope. Variables defined inside a function – inside the function's local scope – are called **local variables**. 

**Variables cannot be both local and global.** So there is a global variable named `number`, and there is a local variable named the same – `number`.

Here's the same code, but with some comments to highlight scopes:


In [441]:
def make_nine():
    # This code block is in a local scope
    number = 9 # Local variable
    return number # The local scope terminates after this line

number = 3 # Global variable

make_nine() # This executes the code within the function, creating a local scope

number # This is still our global variable


3

When `number` is defined in the `make_nine()` function, it only lives in this local scope. When the last line within the function is executed, the local scope is terminated and all variables within gets forgotten.

Here is a picture visualising what was going on:

<img src="../course_material/scopes.png" style="width: 600px;"/>

1. We defined a function called `make_nine()`, and within there is a local variable called `number`
2. We created the global variable `number` – this happend in the global scope
3. We call the function and a local scope is created, and the code block within the function executes
4. The _local_ variable `number` gets assigned to a value – the integer 9 – but this happens within the local scope, _it is a different variable than the global variable `number`_
5. The integer value 9 is returned by the function and the local scope is terminated and the local variable `number` is forgotten
6. Finally, we run the _global_ variable `number` which returns the integer 3 since this is the value the global variable `number` points to

Let's include som `print()` functions to make it even clearer:

In [265]:
def make_nine():
    number = 9 # local variable
    print("Local scope: 'number' is " + str(number))
    return number

In [267]:
number = 3
print("Global scope: 'number' is " + str(number))

# we now call the function, creating a local execution context
make_nine()

print("Global scope: 'number' is still " + str(number) + " in the global scope")

Global scope: 'number' is 3
Local scope: 'number' is 9
Global scope: 'number' is still 3 in the global scope


Hopefully, you now somewhat understand scopes. This concept has some major implications on your code. You will be very confused about bugs in your code if you, for example, don't know which variable lives in which scope. Here are some general rules concerning scopes that will hopefully be useful to remember moving forward:

### 7.1 Global code cannot use local variables

Take this example:

In [455]:
def define_a_variable():
    # This code is in a local scope
    local_num = 42

In [456]:
# calling the function, creating its local scope
define_a_variable()

In [454]:
local_num # error, since there is no such variable in the global scope

NameError: name 'local_num' is not defined

The variable `local_num` doesn't exist in the global scope, it only exists briefly locally when the function is called and the functions code is run. When the function's code block is executed, the local scope terminates and the local variable is forgotten. So when we try to reference that variable in the global scope, the computer has no variable called `local_num` defined in memory. 

### 7.2 Global variables can be used in local scopes

All local scopes can access variables that exists in the global scope. Take this example:

In [464]:
def a_function():
    print(fruit)

In [465]:
fruit = "Banana"

Above, we have a function that only prints a variable named `fruit`. When we call `a_function()`, the interpreter sees that we want to print a variable. But since we havn't created any such variable locally within the function, the interpreter checks to see if any such variable exists in the global scope. So if we run the function:

In [466]:
a_function()

Banana


...it returns the value of the global variable `fruit`.

So the rule is: if there is no `assignement statement` within a local scope, and a variable is referenced, the interpreter will check for any variable with the same name in the global scope and then use any value or object assigned to this name.

### 7.3 Local scopes can't use variables in other local scopes

Let's define two functions:

In [457]:
def function_one():
    big_number = 99
    function_two()
    print(big_number)
    
def function_two():
    big_number = 9000

Here, we have two functions. `function_two()` is called within the scope of `function_one()`. If we call `function_one()`, which number do you think will be printed? Let's have a look:

In [458]:
function_one()

99


When we call `function_one()`, the variable `big_number` is assigned to the value 99. And after that `function_two()` is called, creating its own local scope. Within this scope, `big_number` is assigned to the value 9000 – _but this is another variable than the one in `function_one()`._ They live in different (local) scopes, and are therefore not the same variable.

After `big_number` is defined in `function_two()`, the function is finished and its local scope terminated. The local variable `big_number` in `function_two()` is consequently lost. The interpreter then returns to the local scope within `function_one()` and prints `big_number` – which in _this_ scope is assigned to 99.

The two functions' scopes are not connected and thus, their variables are not interchangable. Let's create a variable in `function_two()` that isn't in `function_one()` and then try to reference it:

In [459]:
def function_one():
    function_two()
    print(fruit)
    
def function_two():
    fruit = "Banana"

In [460]:
function_one()

NameError: name 'fruit' is not defined

We get a `NameError` since the variable `fruit` doesn't exist within the local scope of `function_one()`, only within `function_two()`'s.

### 7.4 Exercise – why won't the code work?

We have this code:

In [46]:
def print_results(results):
    print(f"{str(number1)} multiplied with {str(number2)} is {results}")

def calculator(number1, number2):
    results = number1 * number2

num_apples = 5
num_oranges = 6

calculator(num_apples, num_oranges)
print_results(results)


NameError: name 'results' is not defined

Explain with your own words why the above code doesn't work. Reminder: If you mark a cell in a Jupyter notebook and press <kbd>M</kbd>, the cell will be a text cell (in Markdown) in which you can write down your explaination.

You'll find the solution in the solutions notebook. Good luck!